### ISSUE TITLE
Split big UID range

### ISSUE THREAD

**Issue by eshkinkot** [lefcha/imapfilter#132]
Hello!

I have problem with big UID range. I cannot move messages between mbox, my IMAP server just close my session:

```
S (3): * OK The Microsoft Exchange IMAP4 service is ready.
C (3): 1056 CAPABILITY
S (3): 1056 OK CAPABILITY completed.
C (3): 1057 LOGIN "myusername" *
S (3): 1057 OK LOGIN completed.
C (3): 1058 CAPABILITY
S (3): 1058 OK CAPABILITY completed.
C (3): 1059 NAMESPACE
S (3): 1059 OK NAMESPACE completed.
C (3): 105A SELECT "tmp"
S (3): 105A OK [READ-WRITE] SELECT completed.
C (3): 105B UID COPY 467:28620 "dest"
imapfilter: timeout period expired while waiting to read data
S (3): * OK The Microsoft Exchange IMAP4 service is ready.
C (3): 105C CAPABILITY
S (3): 105C OK CAPABILITY completed.
C (3): 105D LOGIN "myusername" *
S (3): 105D OK LOGIN completed.
C (3): 105E CAPABILITY
S (3): 105E OK CAPABILITY completed.
C (3): 105F NAMESPACE
S (3): 105F OK NAMESPACE completed.
C (3): 1060 SELECT "tmp"
S (3): 1060 OK [READ-WRITE] SELECT completed.
C (3): 1061 UID COPY 467:28620 "dest"
S (3): 1061 NO Server Unavailable. 15
imapfilter: IMAP (3): 1061 NO Server Unavailable. 15
```

So it will be helpful to split big ranges like this `COPY 467:28620` into small ranges.

Please see PR: #133 

With this patch it now looks like this:

```
> res:move_messages(acc.dest)

C (3): 1006 UID COPY 467:517,518:568,569:619,620:670,671:721,722:772,773:823,824:874,875:925,926:976,977:1027,1028:1078,1079:1129,1130:1180,1181:1231,1232:1282,1283:1333,1334:1384,1385:1435,1436:1486,1487:1537,1538:1588,1589:1639,1640:1690,1691:1741,1742:1792,1793:1843,1844:1894,1895:1945,1946:1996,1997:2047,2048:2098,2099:2149,2150:2200,2201:2251,2252:2302,2303:2353,2354:2404,2405:2455,2456:2506,2507:2557,2558:2608,2609:2659,2660:2710,2711:2761,2762:2812,2813:2863,2864:2914,2915:2965,2966:3016 "dest"
S (3): 1006 OK [COPYUID 211 467:3016 289083:291632] COPY completed.
C (3): 1007 UID COPY 3017:3067,3068:3118,3119:3169,3170:3220,3221:3271,3272:3322,3323:3373,3374:3424,3425:3475,3476:3526,3527:3577,3578:3628,3629:3679,3680:3730,3731:3781,3782:3832,3833:3883,3884:3934,3935:3985,3986:4036,4037:4087,4088:4138,4139:4189,4190:4240,4241:4291,4292:4342,4343:4393,4394:4444,4445:4495,4496:4546,4547:4597,4598:4648,4649:4699,4700:4750,4751:4801,4802:4852,4853:4903,4904:4954,4955:5005,5006:5056,5057:5107,5108:5158,5159:5209,5210:5260,5261:5311,5312:5362,5363:5413,5414:5464,5465:5515,5516:5566 "dest"
S (3): 1007 OK [COPYUID 211 3017:5566 291633:294182] COPY completed.
...
C (3): 1012 UID STORE 467:517,518:568,569:619,620:670,671:721,722:772,773:823,824:874,875:925,926:976,977:1027,1028:1078,1079:1129,1130:1180,1181:1231,1232:1282,1283:1333,1334:1384,1385:1435,1436:1486,1487:1537,1538:1588,1589:1639,1640:1690,1691:1741,1742:1792,1793:1843,1844:1894,1895:1945,1946:1996,1997:2047,2048:2098,2099:2149,2150:2200,2201:2251,2252:2302,2303:2353,2354:2404,2405:2455,2456:2506,2507:2557,2558:2608,2609:2659,2660:2710,2711:2761,2762:2812,2813:2863,2864:2914,2915:2965,2966:3016 +FLAGS.SILENT (\Deleted)
S (3): 1012 OK STORE completed.
C (3): 1013 EXPUNGE
S (3): 1013 OK EXPUNGE completed.
C (3): 1014 UID STORE 3017:3067,3068:3118,3119:3169,3170:3220,3221:3271,3272:3322,3323:3373,3374:3424,3425:3475,3476:3526,3527:3577,3578:3628,3629:3679,3680:3730,3731:3781,3782:3832,3833:3883,3884:3934,3935:3985,3986:4036,4037:4087,4088:4138,4139:4189,4190:4240,4241:4291,4292:4342,4343:4393,4394:4444,4445:4495,4496:4546,4547:4597,4598:4648,4649:4699,4700:4750,4751:4801,4802:4852,4853:4903,4904:4954,4955:5005,5006:5056,5057:5107,5108:5158,5159:5209,5210:5260,5261:5311,5312:5362,5363:5413,5414:5464,5465:5515,5516:5566 +FLAGS.SILENT (\Deleted)
S (3): 1014 OK STORE completed.
C (3): 1015 EXPUNGE
S (3): 1015 OK EXPUNGE completed.
```

**Comment by lefcha** [comment]
Interesting, I will look into adding this option to work around such problems.

**Comment by lefcha** [comment]
Did you try setting the `options.limit = 50` or a lower value?  This should break up `STORE` or `COPY` commands that process too many items, into multiple commands.

What does the `range_limit` do different?

**Comment by eshkinkot** [comment]
options.limit break already prepared string with many ranges, if you have
100 000 items and it is sequentially numbered you will have only one range:
UID COPY 1:100000 "dest"
and options.limit does not break it (because string length lower 50).

range_limit break not string with ranges, but range by it self, and you get:
UID COPY 1:50,51:100,101:150, ...,99950:100000 "dest"
and after this, this long string with ranges can be splited by options.limit
now.

I think may be you can rewrite options.limit implementation to do this. I
just try to do minimal code changes (I do not know lua :-)


8 янв. 2017 г. 4:26 PM пользователь "Lefteris Chatzimparmpas" <
notifications@github.com> написал:

Did you try setting the options.limit = 50 or a lower value? This should
break up STORE or COPY commands that process too many items, into multiple
commands.

What does the range_limit do different?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/lefcha/imapfilter/issues/132#issuecomment-271151075>,
or mute the thread

**Comment by lefcha** [comment]
Why is there a need to have such way of breaking up range of message ids? I mean is there some problematic server that doesn't like big ranges?

I'll have a look whether `options.limit` could be enhanced, or a new option like yours is better added.  But let me know if this is a workaround for some problematic scenario...

**Comment by eshkinkot** [comment]
On Sun, Jan 29, 2017 at 5:49 PM, Lefteris Chatzimparmpas <
notifications@github.com> wrote:

> Why is there a need to have such way of breaking up range of message ids?
> I mean is there some problematic server that doesn't like big ranges?
>

Yeah, I have such server at work, I do not control it and do not know how
it configured, it is microsoft exchange.

With big range It is show me timeout:

C (3): 105B UID COPY 467:28620 "dest"
imapfilter: timeout period expired while waiting to read data

or error like:

C (3): 1061 UID COPY 467:28620 "dest"
S (3): 1061 NO Server Unavailable. 15
imapfilter: IMAP (3): 1061 NO Server Unavailable. 15

**Comment by onlyjob** [comment]
I'd like to support this request. I'm trying to use _imapfilter_ to take my INBOX with almost 300_000 messages under control but _imapfilter_ can't move any messages because server timeouts and closes connection on long ranges and `options.limit` does not help... Thanks.

**Comment by ianssoftcom** [comment]
I have had to set `options.limit = 1800` for 50k messages, but I have not had to deal with 300k messages.

**Comment by lefcha** [comment]
I'll add this on the first opportunity, then, seems useful.

**Comment by lefcha** [comment]
Closed by 23a3c625d4bce91c259a6ee18dbac7b2a200e2da.
